### PR TITLE
Add 24-hour clock support

### LINKED ISSUE #682: Feature Request: Support 24-hour / military time
It would be  great to be able to configure the time format from 12:00 to 24:00 hr. So that 11pm is 23:00Hr in Europe. Military time as my wife calls it.... (We are not that military here in Europe tho) :)) And additionally for calendar purposes, to be able to switch the first day of the week from monday to sunday.

### PR DESCRIPTION
**Pull request by ForNeVeR**
I've based my work on #1522, but rewritten everything because there was a strange branch history (guys there have reimported the whole N1 repository from scratch, and merged it with the main branch afterwards) and it was not fully functional.

So I'm the only copyright holder for this contribution, and I've signed CLA, so all the legal issues're covered. Please review this and let me know if any additional changes are required.

It will close #682.

### CODE REVIEW

**Comment by ForNeVeR** [comment]
It seems that the Travis failure has been happened due to some npm problems. Could you please restart it?

**Comment by Cu8eR** [comment]
what about this?
https://github.com/nylas/N1/issues/1917

**Comment by ForNeVeR** [comment]
@Cu8eR, there are *many* date and time formats in the N1 codebase. We either need to define a separate setting for every one of them, or somehow unify them to maybe use single date and time format everywhere. We need to think about it a little bit more, and maybe we need someone more experienced with N1 codebase than me to implement that.

**Comment by bengotow** [comment]
Hey folks! Sorry for the delay on this. I agree there are currently a lot of time/date format strings, and we haven't put enough effort into unifying them. It'd also be really great to avoid adding a setting for this, and instead match the system settings. I was hoping that the new `Intl.DateTimeFormat().resolvedOptions()` in Chromium would allow us to retrieve the user's preference, but it only returns the time zone, numbering system, locale, etc. The `hour12` property is unset...

**Comment by ForNeVeR** [comment]
I agree with you, it would be much better to get the setting from the system-wide configuration instead of messing with N1 settings. Maybe more future-proof solution would be to open an issue at the Chromium bug tracker? Are the guys there know that their implementation lacks of `hour12` option?

File: src/date-utils.es6 [hunk 1]
```diff
@@ -150,6 +150,22 @@ const DateUtils = {
     }
     return moment(date)
   },
+
+  getTimeFormat(opts) {
```
Comment (rpearce) [inline src/date-utils.es6]: If you've got parameter destructuring available to you, I reckon you could clean this code up a bit like so:
```js
getTimeFormat({ seconds, timeZone, upperCase }) {
  const use24HourClock = NylasEnv.config.get('core.workspace.use24HourClock')
  let timeFormat = use24HourClock ? "HH:mm" : "h:mm"
  if (seconds) {
    timeFormat += ':ss'
  }
  if (!use24HourClock && upperCase) {
    timeFormat += ' A'
  }
  if (timeZone) {
    timeFormat += ' z'
  }

  return timeFormat
}
```
While there's still a number of conditionals that could be pulled out and made in to a set of functions that are applied to the original string, but hey why overcomplicate.

Comment (rpearce) [inline src/date-utils.es6]: Sorry for dropping this! Just read the rest of the comments >.<

Comment (ForNeVeR) [inline src/date-utils.es6]: Anyways that's a good suggestion, thank you. I think I'll implement it.

**Comment by Bitbored** [comment]
Any chance this will get merged?

**Comment by bengotow** [comment]
Hey folks! Sorry for the delay getting this merged. This looks really nice—I think inserting just the "time" component of the different format strings based on the setting helped contain the scope of this quite a bit.

Going to flatten and merge this morning! We're working on a major release of N1, so it may be another week before it makes it into a public version. Cheers—

**Comment by bengotow** [comment]
Landed in 899193908ed5ea36dbae8e59757c0181484629f1. Cheers!

**Comment by ForNeVeR** [comment]
Thank you, Ben!

**Pull request closed without merging**